File: python/preprocessing.py
import numpy as np
import mne
from scipy import signal
from sklearn.preprocessing import StandardScaler

def bandpass_filter(raw, l_freq, h_freq):
    """Apply bandpass filter to MNE Raw object."""
    return raw.copy().filter(l_freq, h_freq, fir_design='firwin', verbose=False)

def notch_filter(raw, freqs):
    """Apply notch filter to remove line noise (e.g., 50/60 Hz)."""
    return raw.copy().notch_filter(freqs=freqs, verbose=False)

def run_ica(raw, n_components=15, random_state=42):
    """Run ICA to remove artifacts (e.g., eye blinks, ECG)."""
    ica = mne.preprocessing.ICA(n_components=n_components, random_state=random_state, max_iter='auto')
    ica.fit(raw)
    return ica

def apply_ica(raw, ica, exclude=None):
    """Apply ICA to remove selected components."""
    return ica.apply(raw.copy(), exclude=exclude)

def normalize_signals(data):
    """Normalize signals (numpy array or DataFrame) using z-score."""
    scaler = StandardScaler()
    return scaler.fit_transform(data)

def resample_raw(raw, new_sfreq):
    """Resample MNE Raw object to new sampling frequency."""
    return raw.copy().resample(new_sfreq, npad='auto', verbose=False)

def interpolate_bads(raw, bads):
    """Interpolate bad channels in MNE Raw object."""
    raw.info['bads'] = bads
    return raw.copy().interpolate_bads(reset_bads=True, verbose=False)

File: R/preprocessing.R
library(signal)
library(fastICA)

bandpass_filter <- function(signal, low, high, fs) {
  bf <- butter(4, c(low, high) / (fs / 2), type = "pass")
  filtfilt(bf, signal)
}

notch_filter <- function(signal, freq, fs) {
  w0 <- freq / (fs / 2)
  bw <- w0 / 35
  nf <- signal::butter(2, c(w0 - bw, w0 + bw), type = "stop")
  filtfilt(nf, signal)
}

run_ica <- function(data, n.comp = 15) {
  fastICA::fastICA(data, n.comp)
}

apply_ica <- function(data, ica, exclude = NULL) {
  if (!is.null(exclude)) {
    ica$S[, exclude] <- 0
    return(ica$A %*% t(ica$S))
  }
  return(ica$A %*% t(ica$S))
}

normalize_signals <- function(data) {
  scale(data)
}

resample_signal <- function(signal, orig_fs, new_fs) {
  n <- length(signal)
  t_orig <- seq(0, (n - 1) / orig_fs, by = 1 / orig_fs)
  t_new <- seq(0, max(t_orig), by = 1 / new_fs)
  approx(t_orig, signal, xout = t_new)$y
}

interpolate_bads <- function(data, bads) {
  for (bad in bads) {
    data[, bad] <- zoo::na.approx(data[, bad], na.rm = FALSE)
  }
  return(data)
}
